File: sdks/py_sdk/sw4rm/clients/connector.py
from __future__ import annotations

from typing import Any


class ConnectorClient:
    """Client for interacting with the ConnectorService.

    The ConnectorService manages tool providers and their capabilities.
    """

    def __init__(self, channel: Any, timeout: float = 30.0) -> None:
        """Initialize the ConnectorClient.

        Args:
            channel: gRPC channel to use for communication
            timeout: Default timeout in seconds for RPC calls (default: 30.0)
        """
        self._channel = channel
        self._timeout = timeout
        try:
            from sw4rm.protos import connector_pb2, connector_pb2_grpc  # type: ignore
            self._pb2 = connector_pb2
            self._stub = connector_pb2_grpc.ConnectorServiceStub(channel)
        except Exception:
            self._pb2 = None
            self._stub = None

    def register_provider(
        self,
        provider_id: str,
        tools: list[dict[str, Any]],
        timeout: float | None = None
    ) -> Any:
        """Register a tool provider with its available tools.

        Args:
            provider_id: Unique identifier for the provider
            tools: List of tool descriptors, each containing:
                - tool_name (str): Name of the tool
                - input_schema (str): JSON Schema or URL for input
                - output_schema (str): JSON Schema or URL for output
                - idempotent (bool): Whether the tool is idempotent
                - needs_worktree (bool): Whether the tool needs a worktree
                - default_timeout_s (int): Default timeout in seconds
                - max_concurrency (int): Maximum concurrent executions
                - side_effects (str): Side effects like "filesystem", "network"
            timeout: Optional timeout in seconds (uses default if not specified)

        Returns:
            ProviderRegisterResponse with ok (bool) and reason (str) fields

        Raises:
            RuntimeError: If protobuf stubs are not generated
        """
        if not self._stub or not self._pb2:
            raise RuntimeError("Protobuf stubs not generated. Run `make protos`.")

        tool_descriptors = [
            self._pb2.ToolDescriptor(**tool) for tool in tools
        ]
        req = self._pb2.ProviderRegisterRequest(
            provider_id=provider_id,
            tools=tool_descriptors
        )
        return self._stub.RegisterProvider(req, timeout=timeout or self._timeout)

    def describe_tools(self, provider_id: str, timeout: float | None = None) -> Any:
        """Describe the tools available from a specific provider.

        Args:
            provider_id: Unique identifier for the provider
            timeout: Optional timeout in seconds (uses default if not specified)

        Returns:
            DescribeToolsResponse with a list of ToolDescriptor objects

        Raises:
            RuntimeError: If protobuf stubs are not generated
        """
        if not self._stub or not self._pb2:
            raise RuntimeError("Protobuf stubs not generated. Run `make protos`.")
        req = self._pb2.DescribeToolsRequest(provider_id=provider_id)
        return self._stub.DescribeTools(req, timeout=timeout or self._timeout)

File: sdks/js_sdk/src/clients/connector.ts
import * as grpc from '@grpc/grpc-js';
import { BaseClient, ClientOptions } from '../internal/baseClient.js';

type ConnectorServiceClient = grpc.Client & {
  RegisterProvider(req: ProviderRegisterRequest, meta: grpc.Metadata, opts: grpc.CallOptions, cb: (err: grpc.ServiceError | null, res: ProviderRegisterResponse) => void): void;
  DescribeTools(req: DescribeToolsRequest, meta: grpc.Metadata, opts: grpc.CallOptions, cb: (err: grpc.ServiceError | null, res: DescribeToolsResponse) => void): void;
};

export interface ToolDescriptor {
  tool_name: string;
  input_schema?: string;
  output_schema?: string;
  idempotent?: boolean;
  needs_worktree?: boolean;
  default_timeout_s?: number;
  max_concurrency?: number;
  side_effects?: string;
}

export interface ProviderRegisterRequest { provider_id: string; tools: ToolDescriptor[] }
export interface ProviderRegisterResponse { ok: boolean; reason?: string }
export interface DescribeToolsRequest { provider_id: string }
export interface DescribeToolsResponse { tools: ToolDescriptor[] }

export class ConnectorClient extends BaseClient {
  private client: ConnectorServiceClient;
  constructor(opts: ClientOptions) {
    super(opts);
    this.client = this.getServiceClient<ConnectorServiceClient>('sw4rm.connector.ConnectorService');
  }

  async registerProvider(req: ProviderRegisterRequest): Promise<ProviderRegisterResponse> {
    const meta = this.metadata();
    const opts: grpc.CallOptions = { deadline: this.deadlineFromNow() };
    return this.withRetryUnary(() => new Promise<ProviderRegisterResponse>((resolve, reject) => {
      this.client.RegisterProvider(req, meta, opts, (err, res) => {
        if (err) return reject(err);
        resolve(res);
      });
    }));
  }

  async describeTools(providerId: string): Promise<ToolDescriptor[]> {
    const meta = this.metadata();
    const opts: grpc.CallOptions = { deadline: this.deadlineFromNow() };
    return this.withRetryUnary(() => new Promise<ToolDescriptor[]>((resolve, reject) => {
      this.client.DescribeTools({ provider_id: providerId }, meta, opts, (err, res) => {
        if (err) return reject(err);
        resolve(res?.tools ?? []);
      });
    }));
  }
}

